### ISSUE TITLE
async-safe way of reading this.state?

### ISSUE THREAD

**Issue by linkenneth** [facebook/react#12381]
**Do you want to request a *feature* or report a *bug*?**
feature request / question?

I'm having a problem that I'm sure some people have had, though I feel like it should be much more solvable. Suppose I have a component that receives state updates from many sources. For example,

```javascript
class Counter extends React.Component {
  constructor (props) {
    this.state = { number : 0 };
  }

  onButtonAClicked () {
    this.setState((prevState) => { prevState.number + 1 });
  }

  onButtonBClicked () {
    this.setState((prevState) => { prevState.number + 2 });
  }
}
```

Now, somewhere outside of the `render()` method I need to read `this.state`. Since `this.setState` is asynchronous, there is no safe way to read this number if you don't know the sequence of clicks since the last actual update of this.state. One safe way seems to be using an empty `setState` with a callback to do things:

```javascript
readAndDoStuff () {
  this.setState({}, () => {
    // actually do things with this.state
  });
}
```

but my requirement is that this method that reads this.state must be synchronous. Also, this form seems a bit odd to me if I don't actually intend to set any state. It seems that `this.setState` being asynchronous forcing my own application to have to make all reads of `this.state` to be asynchronous as well.

I could instead keep track of my own state and entirely ignore `this.state`. This means I keep `this.state2 = { number: 1 };` and replace calls of `this.setState` with `this.state2[number]++; this.forceUpdate();`, but this just solves the problem by bypassing React entirely.

My question is: is there a way to force the immediate setting of state to read `this.state`, or what is a recommended solution for something like this?

**Comment by gaearon** [comment]
Can you explain more about the use case? Why do you need to read the pending state rather than the one the user sees on the screen?

**Comment by linkenneth** [comment]
Thanks for the quick response. I'm implementing a game. In the longer-term I intend to render the game with something more complex, but in the meantime I've just been wrapping it in a React component. This problem appears to be more general than the particularities of the game though...

The state of the world is represented by a grid: (W represents wall, > represents the player's character, . represents the ground etc.)

```
. . . W .
. > . W .
. . . . .
```

The `this.state` of the world can roughly be thought of as
```javascript
this.state = {
  position: [1, 1],
  orientation: EAST
}
```

The game is a bit special in that players can write scripts to execute a sequence of actions in order. However, the problem is more general than this game makes it appear. Whether through the players' scripts or otherwise, it is necessary to be able to perform checks such as these:

```javascript
move()
if (frontIsBlocked()) {
  // do stuff
}
```

We must able to correctly return `frontIsBlocked()` on the new state if the player writes such a script. Now I know this seems very particular to my game, but it seems that being able to handle state transitions at any speed without having to know all the actions that cause these transitions ahead of time is something that should be able to be solved in React? At the point of calling `frontIsBlocked()` to read the world's state, I shouldn't have to know that I just `move()`'d three times and turned left, right? It just seems weird to me that any calls to `this.setState` means that we lose consistency in `this.state` for an unknown amount of time, and there's no method to "force" data to be consistent again before reading. This means that we can only call `this.setState` in "one go", where we know all the changes that will happen to it (?)?

Please let me know if I'm missing something critical! Thanks!

**Comment by gaearon** [comment]
Where is `move` called from? If it’s a result of calling `setState` then you’ll get the previous state as an argument of the updater function. Can you explain more about why this is not sufficient?

**Comment by gaearon** [comment]
In other words

```js
runPlayerScriptsForTick() {
  this.setState(runScripts);
} 

// ...
function runScripts(state) {
  // run all scripts and return the next state
}
```

**Comment by linkenneth** [comment]
Thanks again for the response! Hmm, I think this will work but requires some custom transformations of the player's commands to correctly execute in that context and use state. I'll keep this solution in mind as I implement, but right now this seems a bit fragile, so I'm leaning towards keeping state outside of `this.state` instead.

I'm happy to close this issue as there are a few solutions, but what I really wanted to discuss was the higher-level issue of whether we can easily force consistency of `this.state`? Is this an issue that has ever come up before for you guys? A reference to a Promise that resolves when the state is consistent, or a `this.forceConsistent()` would all be incredibly helpful to my situation.

Here's something I found that's similar: https://stackoverflow.com/questions/41896878/react-setstate-getstate-and-asynchronous

**Comment by gaearon** [comment]
>Hmm, I think this will work but requires some custom transformations of the player's commands to correctly execute in that context and use state

This is a bit hard to discuss without a specific full example.

You can force `setState` to be synchronous:

```js
ReactDOM.flushSync(() => {
  this.setState(...)
});
```

That’s a bad perf practice though.
There are good reasons why React doesn’t flush state immediately.

I encourage you to check these resources out where I went in depth on it:

https://stackoverflow.com/questions/48563650/does-react-keep-the-order-for-state-updates/48610973#48610973

https://github.com/facebook/react/issues/11527#issuecomment-360199710

I’m happy to continue this discussion with a full code example (preferably not too long) to anchor it.

**Comment by linkenneth** [comment]
Thanks! Let me take a look at those resources and see whether these work. The performance hit (haven't tested yet) may not be that bad, since we might want to execute the player's commands one line at a time anyways to "animate" the moves. Do you mind if I continue implementing and get back to you?

**Comment by gaearon** [comment]
Sure

**Comment by linkenneth** [comment]
I was able to solve my issue via other means so closing this for now.

For the record, I realized that for my specific use case it was necessary to separate the "rendered" state of the world and the "calculated" state of the world. Player scripts should be able to run immediately on the "calculated" state of the world, but the "rendered" state of the world should be updated at a delay afterwards. Therefore, I decided to keep two states of the world, and update the "rendered" state (ie. `this.state`) in order to "animate" these state changes at a delay. I also kept a history of the intermediate state changes necessary to reach the "calculated" state from the "rendered" state.

I'm unsure if this is the best way to approach this, but it seems reasonable to me since we do indeed need these two types of state to be different at various times. Thanks for the support!
